File: packages/pdp-geo-generator-agent/src/rag/index-skeleton.ts
import { chunkPdpGeoRagDocument } from "./retrieval";
import { findPdpGeoRagIndexEntry, findPdpGeoRagSectionEntry } from "./rag-index";
import type { PdpGeoRagFieldTarget, PdpGeoRagIntent, PdpGeoRagKind } from "../types";

/**
 * Deterministic rag-index skeleton generator (PageIndex/RAPTOR-style tree
 * indexing, offline half only). It parses the real heading tree of every
 * corpus document, reports which headings already resolve to a typed
 * `rag-index.ts` section, and emits ready-to-paste section entries (with
 * heuristically inferred intents/fieldTargets) for the ones that do not.
 * No LLM and no runtime cost: the output is reviewed by a human and committed,
 * keeping retrieval deterministic while removing manual index drift.
 */
export interface PdpGeoRagIndexSkeletonSection {
  heading: string;
  headingPath: string;
  indexed: boolean;
  /** Routing the runtime would infer today (index entry when indexed, heuristic otherwise). */
  intents: PdpGeoRagIntent[];
  fieldTargets: PdpGeoRagFieldTarget[];
}

export interface PdpGeoRagIndexSkeletonEntry {
  document: string;
  version: string;
  kind: PdpGeoRagKind;
  documentIndexed: boolean;
  sections: PdpGeoRagIndexSkeletonSection[];
  unindexedSectionCount: number;
}

export function createPdpGeoRagIndexSkeleton(
  documents: Array<{ name: string; content: string; version?: string }>
): PdpGeoRagIndexSkeletonEntry[] {
  return documents.map((document) => {
    const chunks = chunkPdpGeoRagDocument(document.name, document.content, document.version ?? "v1");
    const indexEntry = findPdpGeoRagIndexEntry(document.name);
    const seen = new Set<string>();
    const sections: PdpGeoRagIndexSkeletonSection[] = [];

    for (const chunk of chunks) {
      const heading = chunk.title ?? "";
      const headingPath = typeof chunk.metadata.headingPath === "string" ? chunk.metadata.headingPath : heading;
      const key = `${heading}::${headingPath}`;
      if (!heading || seen.has(key)) {
        continue;
      }
      seen.add(key);
      sections.push({
        heading,
        headingPath,
        indexed: findPdpGeoRagSectionEntry(document.name, heading, headingPath) !== undefined,
        intents: chunk.intents ?? [],
        fieldTargets: chunk.fieldTargets ?? []
      });
    }

    return {
      document: document.name,
      version: document.version ?? "v1",
      kind: chunks[0]?.kind ?? "custom",
      documentIndexed: indexEntry !== undefined,
      sections,
      unindexedSectionCount: sections.filter((section) => !section.indexed).length
    };
  });
}

/** Renders unindexed headings as paste-ready `rag-index.ts` section entries. */
export function renderPdpGeoRagIndexSkeleton(entries: PdpGeoRagIndexSkeletonEntry[]): string {
  const lines: string[] = [];

  for (const entry of entries) {
    const unindexed = entry.sections.filter((section) => !section.indexed);
    if (!entry.documentIndexed) {
      lines.push(`// MISSING DOCUMENT ENTRY: ${entry.document} (kind: ${entry.kind}, version: ${entry.version})`);
    }
    if (unindexed.length === 0) {
      continue;
    }
    lines.push(`// ${entry.document}: ${unindexed.length} unindexed heading(s)`);
    for (const section of unindexed) {
      lines.push([
        "{",
        `  heading: ${JSON.stringify(section.heading)},`,
        `  intents: ${JSON.stringify(section.intents)},`,
        `  fieldTargets: ${JSON.stringify(section.fieldTargets)},`,
        "  priority: 0.8",
        "},"
      ].join("\n"));
    }
  }

  return lines.join("\n");
}

File: packages/pdp-geo-generator-agent/src/pdp_geo_generator_agent/rag/index_skeleton.py
"""Offline typed-index skeleton generator for reviewable RAG maintenance."""

from __future__ import annotations

import json
from collections.abc import Mapping, Sequence
from typing import Any

from .._json import as_dict, as_list
from .index import find_pdp_geo_rag_index_entry, find_pdp_geo_rag_section_entry
from .retrieval import chunk_pdp_geo_rag_document


def create_pdp_geo_rag_index_skeleton(documents: Sequence[Mapping[str, Any]]) -> list[dict[str, Any]]:
    entries: list[dict[str, Any]] = []
    for document in documents:
        name, version = str(document.get("name") or ""), str(document.get("version") or "v1")
        chunks = chunk_pdp_geo_rag_document(name, str(document.get("content") or ""), version)
        index = find_pdp_geo_rag_index_entry(name)
        sections: list[dict[str, Any]] = []
        seen: set[str] = set()
        for chunk in chunks:
            heading = str(chunk.get("title") or "")
            metadata = as_dict(chunk.get("metadata"))
            path = str(metadata.get("headingPath") or heading)
            key = f"{heading}::{path}"
            if not heading or key in seen:
                continue
            seen.add(key)
            sections.append(
                {
                    "heading": heading,
                    "headingPath": path,
                    "indexed": find_pdp_geo_rag_section_entry(name, heading, path) is not None,
                    "intents": list(chunk.get("intents") or []),
                    "fieldTargets": list(chunk.get("fieldTargets") or []),
                }
            )
        entries.append(
            {
                "document": name,
                "version": version,
                "kind": chunks[0].get("kind", "custom") if chunks else "custom",
                "documentIndexed": index is not None,
                "sections": sections,
                "unindexedSectionCount": sum(not section["indexed"] for section in sections),
            }
        )
    return entries


def render_pdp_geo_rag_index_skeleton(entries: Sequence[Mapping[str, Any]]) -> str:
    lines: list[str] = []
    for raw in entries:
        entry = as_dict(raw)
        unindexed = [as_dict(item) for item in as_list(entry.get("sections")) if not as_dict(item).get("indexed")]
        if not entry.get("documentIndexed"):
            lines.append(
                f"// MISSING DOCUMENT ENTRY: {entry.get('document', '')} (kind: {entry.get('kind', 'custom')}, version: {entry.get('version', 'v1')})"
            )
        if not unindexed:
            continue
        lines.append(f"// {entry.get('document', '')}: {len(unindexed)} unindexed heading(s)")
        for section in unindexed:
            lines.extend(
                [
                    "{",
                    f"  heading: {json.dumps(section.get('heading', ''), ensure_ascii=False)},",
                    f"  intents: {json.dumps(section.get('intents', []), ensure_ascii=False)},",
                    f"  fieldTargets: {json.dumps(section.get('fieldTargets', []), ensure_ascii=False)},",
                    "  priority: 0.8",
                    "},",
                ]
            )
    return "\n".join(lines)


createPdpGeoRagIndexSkeleton = create_pdp_geo_rag_index_skeleton
renderPdpGeoRagIndexSkeleton = render_pdp_geo_rag_index_skeleton
